### PR TITLE
Add microscopy tutorial

### PR DESCRIPTION
**Pull request by mariehbourget**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [ ] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
This PR aims to add a two-classes microscopy tutorials to the documentation and do minor updates to the microscopy config file.

## Linked issues
No related issue.

### CODE REVIEW

**Comment by mariehbourget** [comment]
This PR for the new microscopy training tutorial is ready for review.

The figure included in the tutorial is from PR https://github.com/ivadomed/doc-figures/pull/5 which need to be approved and merged before this one, so I can update the link to the figure.

@mathieuboudreau, I'm not able to tag you as reviewer but you may be interested in this tutorial as well.

File: docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst [hunk 1]
```diff
@@ -0,0 +1,305 @@
+Two-classes microscopy segmentation with 2D U-Net
```
Comment (jcohenadad) [inline docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst:1]:
```suggestion
Two-class microscopy segmentation with 2D U-Net
```

File: docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+- Training of a segmentation model (U-Net 2D) with a two classes labels on a single contrast on microscopy PNG images,
+
+- Testing of a trained model and computation of evaluation metrics,
+
+- Visualization of the outputs of a trained model.
+
+Download dataset
+----------------
+
+We will use a publicly-available dataset consisting of 10 microscopy sample of rat spinal cord.
```
Comment (jcohenadad) [inline docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst:15]:
```suggestion
We will use a publicly-available dataset consisting of 10 microscopy samples of rat spinal cord.
```

File: docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+     "path_output": "log_microscopy_sem"
+
+- ``loader_parameters:path_data``: Location of the dataset. As discussed in `Data <../data.html>`__, the dataset
+  should conform to the BIDS standard. Modify the path so it points to the location of the downloaded dataset.
+
+  .. code-block:: xml
+
+     "path_data": "data_axondeepseg_sem"
+
+- ``loader_parameters:bids_config``: Location of the custom bids configuration file required for microscopy
```
Comment (jcohenadad) [inline docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst:59]:
```suggestion
- ``loader_parameters:bids_config``: Location of the custom BIDS configuration file required for microscopy
```

File: docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+  file indexing.
+
+  .. code-block:: xml
+
+     "bids_config": "ivadomed/config/config_bids.json",
+
+- ``loader_parameters:target_suffix``: Suffix of the ground truths segmentation. The ground truths are located
```
Comment (jcohenadad) [inline docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst:66]:
```suggestion
- ``loader_parameters:target_suffix``: Suffix of the ground truth segmentations. The ground truths are located
```

File: docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+   "command": "test"
+
+Then run:
+
+.. code-block:: bash
+
+   ivadomed -c config_microscopy.json
+
+The model's parameters will be displayed in the terminal, followed by a preview of the results for each image.
+The resulting segmentations are saved for each image in the ``<PATH_TO_OUT_DIR>/pred_masks`` while a csv file,
```
Comment (jcohenadad) [inline docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst:244]:
```suggestion
The resulting segmentations are saved for each image in the ``<PATH_TO_OUT_DIR>/pred_masks`` while a CSV file,
```

File: docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst [hunk 6]
```diff
@@ -0,0 +1,305 @@
+Two-class microscopy segmentation with 2D U-Net
+=================================================
```
Comment (jcohenadad) [inline docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst:2]:
```suggestion
===============================================
```

File: docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+
+In this tutorial we will learn the following features:
+
+- Training of a segmentation model (U-Net 2D) with a two classes labels on a single contrast on microscopy PNG images,
```
Comment (jcohenadad) [inline docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst:6]:
```suggestion
- Training of a segmentation model (U-Net 2D) with two-class labels on a single contrast on microscopy PNG images,
```

File: docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+
+- Testing of a trained model and computation of evaluation metrics,
+
+- Visualization of the outputs of a trained model.
+
+Download dataset
+----------------
+
+We will use a publicly-available dataset consisting of 10 microscopy samples of rat spinal cord.
```
Comment (jcohenadad) [inline docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst:15]:
```suggestion
We will use a publicly available dataset consisting of 10 microscopy samples of rat spinal cord.
```

File: docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+  between 256x256 and 512x512. In our case, we use patches of 256x256.
+- ``default_model:stride_2D``: Pixels’ shift over the input matrix to create 2D patches. In our case, we use
+  a stride of 244 pixels in both dimensions, resulting in an overlap of 12 pixels between patches.
+
+  .. code-block:: xml
+
+     "length_2D": [256, 256]
+     "stride_2D": [244, 244]
+
+- ``postprocessing:binarize_maxpooling``: Use to binarize predictions across all classes in multiclass models.
```
Comment (jcohenadad) [inline docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst:128]:
```suggestion
- ``postprocessing:binarize_maxpooling``: Used to binarize predictions across all classes in multiclass models.
```

File: docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+
+  .. code-block:: xml
+
+      "binarize_maxpooling": {}
+
+- ``transformation:Resample``: Use to resample images to a common resolution (in mm) before splitting into patches,
```
Comment (jcohenadad) [inline docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst:134]:
```suggestion
- ``transformation:Resample``: Used to resample images to a common resolution (in mm) before splitting into patches,
```

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+- ``split_dataset:test_fraction``: Fraction of the dataset's ``sample_id`` in the test set. In our case ``0.1``.
+
+  .. code-block:: xml
+
+      "split_method" : "sample_id"
+      "train_fraction": 0.6
+      "test_fraction": 0.1
+
+- ``training_time:num_epochs``: The maximum number of epochs that will be run during training. Each epoch is composed
+  of a training part and an evaluation part. It should be a strictly positive integer. In our case, we will use
```
Comment (andreanne-lemay) [inline docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst:111]: This is just a suggestion as to me evaluation sounds more on the test set and validation on the validation set, but please ignore if you disagree. Later in the tutorial, there is a section named `Evaluate model` which refers to inference on the testing set. Using evaluation here might be confusing.

```suggestion
  of a training part and a validation part. It should be a strictly positive integer. In our case, we will use
```

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+- ``split_dataset:train_fraction``: Fraction of the dataset's ``sample_id`` in the train set. In our case ``0.6``.
+- ``split_dataset:test_fraction``: Fraction of the dataset's ``sample_id`` in the test set. In our case ``0.1``.
+
+  .. code-block:: xml
+
+      "split_method" : "sample_id"
+      "train_fraction": 0.6
+      "test_fraction": 0.1
+
+- ``training_time:num_epochs``: The maximum number of epochs that will be run during training. Each epoch is composed
```
Comment (andreanne-lemay) [inline docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst:110]:
```suggestion
- ``training_parameters:training_time:num_epochs``: The maximum number of epochs that will be run during training. Each epoch is composed
```

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+  between 256x256 and 512x512. In our case, we use patches of 256x256.
+- ``default_model:stride_2D``: Pixels’ shift over the input matrix to create 2D patches. In our case, we use
+  a stride of 244 pixels in both dimensions, resulting in an overlap of 12 pixels between patches.
+
+  .. code-block:: xml
+
+     "length_2D": [256, 256]
+     "stride_2D": [244, 244]
+
+- ``postprocessing:binarize_maxpooling``: Used to binarize predictions across all classes in multiclass models.
```
Comment (andreanne-lemay) [inline docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst:128]:
```suggestion
- ``postprocessing:binarize_maxpooling``: Used to binarize predictions across all classes in multiclass models. For each pixel, the class, including the background class, with the highest output probability will be segmented.
```

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+
+  .. code-block:: xml
+
+      "binarize_maxpooling": {}
+
+- ``transformation:Resample``: Used to resample images to a common resolution (in mm) before splitting into patches,
+  according to each image real pixel size. In our case, we resample the images to a common resolution of 0.0001 mm
+  (0.1 um) in both dimensions.
```
Comment (andreanne-lemay) [inline docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst:136]:
```suggestion
  (0.1 μm) in both dimensions.
```

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+.. note::
+
+   If a `compatible GPU <https://pytorch.org/get-started/locally/>`_ is available, it will be used by default.
+   Otherwise, training will use the CPU, which will take a prohibitively long computational time (several hours).
+
+The main parameters of the training scheme and model will be displayed on the terminal, followed by the loss value
+on training and validation sets at every epoch. To know more about the meaning of each parameter, go to
+:doc:`../configuration_file`. The value of the loss should decrease during the training.
+
+.. code-block:: console
```
Comment (andreanne-lemay) [inline docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst:172]: We should probably update all our tutorials eventually (not the most important thing), but the output in the console doesn't match the output in our tutorials (regardless of the INFO part, some prints don't match): 
image

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+      "train_fraction": 0.6
+      "test_fraction": 0.1
+
+- ``training_time:num_epochs``: The maximum number of epochs that will be run during training. Each epoch is composed
+  of a training part and an evaluation part. It should be a strictly positive integer. In our case, we will use
+  50 epochs.
+
+  .. code-block:: xml
+
+     "num_epochs": 50
```
Comment (andreanne-lemay) [inline docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst:116]: After ~15 epochs I reach ~85% of Dice score on the validation set. We could consider reducing the number of epochs as it's only for a tutorial.

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+   Initialising model's weights from scratch.
+
+   Scheduler parameters: {'name': 'CosineAnnealingLR', 'base_lr': 1e-05, 'max_lr': 0.01}
+
+   Selected Loss: DiceLoss
+   with the parameters: []
+   Epoch 1 training loss: -0.6894.
+   Epoch 1 validation loss: -0.7908.
+
+After 50 epochs (see ``"num_epochs"`` in the configuration file), the Dice score on the validation set should be ~85%.
```
Comment (andreanne-lemay) [inline docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst:215]: Same comment as above, after 15 epochs I reach this performance.

**Review by andreanne-lemay — APPROVED** [review]
I added minor suggestion changes (please ignore where you see fit). Otherwise, it's good to go for me, great work!

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+  .. code-block:: xml
+
+     "path_output": "log_microscopy_sem"
+
+- ``loader_parameters:path_data``: Location of the dataset. As discussed in `Data <../data.html>`__, the dataset
+  should conform to the BIDS standard. Modify the path so it points to the location of the downloaded dataset.
+
+  .. code-block:: xml
+
+     "path_data": "data_axondeepseg_sem"
```
[the remark this answers is not in the document]
Comment (mariehbourget) [reply docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst]: Good catch! Resolved in e075dee.

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst (hunk 11 above)
Comment (mariehbourget) [reply docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst]: Agreed!

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst (hunk 15 above)
Comment (mariehbourget) [reply docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst:172]: It may be a mistake on my part when removing the "date/info" part of the output. I'll double-check.

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst (hunk 16 above)
Comment (mariehbourget) [reply docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst:116]: Interesting!
I'll do a training with only 15-20 epochs to see if the segmentation look similar to the segmentations after 50 epochs (I had some surprising results in the past even with similar Dice).
If I can, I'll reduce the number of epochs and update the figure. Thanks!

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst (hunk 15 above)
Comment (mariehbourget) [reply docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst:172]: I did not find significant differences but made some minor aesthetic changes in 69b2901.

File: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst (hunk 16 above)
Comment (mariehbourget) [reply docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst:116]: As I was suspecting, even with a similar Dice after 20 epochs, the segmentation results are not as good (a lot of noise and false positive in the background). I'm not sure why this happens... maybe because the reported Dice after each epochs is a mean of all classes?
Anyway, for now I would prefer to keep the 50 epochs so we have good segmentations like in the figure.

Comment (andreanne-lemay) [reply docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst:116]: Sounds good to me :)!

**Comment by andreanne-lemay** [comment]
> Thank you @andreanne-lemay for the review, I made most of the requested changes but kept 50 epochs to ensure a good segmentation results and match the figure. If that works for you, would you be able to approve this PR [ivadomed/doc-figures#5](https://github.com/ivadomed/doc-figures/pull/5)? (It needs to go first so I can update the link to the figure). Thanks!

Done :heavy_check_mark:  Thanks to you!
